provider/kubernetes/kubernetes.go

```diff
@@ -454,18 +458,23 @@ func getTLS(ingress *extensionsv1beta1.Ingress, k8sClient Client) ([]*tls.Config
 	return tlsConfigs, nil
 }
 
-func endpointPortNumber(servicePort corev1.ServicePort, endpointPorts []corev1.EndpointPort) int {
-	if len(endpointPorts) > 0 {
-		// name is optional if there is only one port
-		port := endpointPorts[0]
-		for _, endpointPort := range endpointPorts {
-			if servicePort.Name == endpointPort.Name {
-				port = endpointPort
-			}
+// endpointPortNumber returns the port to be used for this endpoint. It is zero
+// if the endpoint does not match the given service port.
+func endpointPortNumber(servicePort corev1.ServicePort, endpointPorts []corev1.EndpointPort) int32 {
+	// Is this reasonable to assume?
+	if len(endpointPorts) == 0 {
+		return servicePort.Port
+	}
+
+	for _, endpointPort := range endpointPorts {
+		// For matching endpoints, the port names must correspond, either by
+		// being empty or non-empty. Multi-port services mandate non-empty
+		// names and allow us to filter for the right addresses.
+		if servicePort.Name == endpointPort.Name {
```

Comment:
Looking at the docs: (https://godoc.org/k8s.io/api/core/v1#EndpointPort) It appears that `endpointPort.Name` is optional if only one port is defined. We may have to go deeper, to see if `servicePort.Port` == `endpointPort.Port` too. I can't think of a scenario of the top of my head where this would matter, but in theory it could (where an ingress uses a numeric port, and the endpoint is named?)

Reply:
You are right that the endpoints port is optional if and only if a single service port is defined. ([Reference](https://v1-9.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#endpointport-v1-core).)

I don't think that the associated ports would ever differ then. Regarding your example, is it really relevant? The ports that an Endpoints object exposes is defined by the Service, not the Ingress.